### PR TITLE
fix(datepickers): stop calling onChange from within DatePicker reducer

### LINKED ISSUE #1359: Datepicker component gives "Cannot update a component while rendering a different component (Datepicker)" warning
## Expectations

When following the example in the [Datepicker](https://garden.zendesk.com/components/date-picker) documentation, the code should work without errors/warnings.

## Reality

The first time the state is updated, this warning appears.
image

## Steps to Reproduce

1.Use the code in the documented example inside of another component.
2.Select any date

Codesandbox: https://codesandbox.io/s/agitated-silence-vb9euf

### Fine Print

- Component: `Datepicker`
- Version: 8.52.0
- Browsers: Chrome

### PR DESCRIPTION
**Pull request by jgorfine-zendesk**
## Description

**Resolves #1359**

`DatePicker` triggered a "Cannot update a component while rendering a different component" console warning because `datepickerReducer` called the `onChange` prop as a side effect inside the reducer body (for both `SELECT_DATE` and `MANUALLY_UPDATE_INPUT`). React re-invokes reducers during the render phase to process dispatched actions, so that `onChange` call — and the consumer's `setState` it triggers — fired while `DatePicker` was mid-render.

This surfaced in the CodeSandbox repro and not in this repo's own Storybook, because Storybook's Controls addon manages args via its own store (`useArgs`/`updateArgs`), not a real `setState` in the render tree — so the default `DatePicker` stories never actually exercised a controlled `onChange` consumer the way a real app (or the CodeSandbox prototype) would. The jest repro below reproduces it directly against a controlled `useState`/`onChange` consumer, matching the original report.

`onChange` is now called directly from the real event handlers (`Calendar`'s day `onClick`, `Input`'s `handleChange`) instead of from the reducer, matching the pattern `DatePickerRange`'s `Start`/`End`/`Month` already use. The reducer is now pure, with no side effects.

Also added a permanent "Custom date format" pattern story (`[patterns]` group) demonstrating `formatDate`/`customParseDate` as a matched pair, since neither prop had a live documented example before, and both needed verification after being threaded through as explicit props instead of reducer-internal state.

https://github.com/user-attachments/assets/ccf81849-5067-4980-a454-51aea89dbdca

## Detail

Reproduced locally with a jest spy on `console.error`, mounting a controlled `DatePicker` (`useState`/`onChange={setState}`) matching the original CodeSandbox repro. The stack trace pointed directly at the bug:

```
onChange (date-picker-reducer.ts:153:11)
  at updateReducer (react-dom...)
  at useReducer (DatePicker.tsx:61:39)
  at renderWithHooks ...
```

Also manually verified in a real browser (Storybook) that:
- `onChange` still fires correctly via both `Calendar` (day click) and `Input` (typed date)
- the `value`-based dedupe (no duplicate `onChange` on re-selecting the same date) still works
- `formatDate`/`customParseDate` still round-trip correctly (new pattern story: `Packages/DatePickers/[patterns]` → "Custom date format")

## Checklist

- ~~:ok_hand: design updates will be Garden Designer approved (add the designer as a reviewer)~~
- [x] :globe_with_meridians: demo is up-to-date (`npm start`)
- ~~:arrow_left: renders as expected with reversed (RTL) direction~~
- ~~:black_circle: renders as expected in dark mode~~
- ~~:metal: renders as expected with Bedrock CSS (`?bedrock`)~~
- [ ] :guardsman: includes new unit tests. Maintain existing coverage (always >= 96%)
- ~~:wheelchair: tested for WCAG 2.1 AA accessibility compliance~~
- [ ] :memo: tested in Chrome, Firefox, Safari, and Edge

---

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/datepickers/src/elements/DatePicker/components/Input.tsx [hunk 1]
```diff
@@ -6,26 +6,38 @@
...
       dispatch({ type: 'CLOSE' });
     };
 
     const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
-      dispatch({ type: 'MANUALLY_UPDATE_INPUT', value: e.target.value });
+      const inputValue = e.target.value;
+      const currentDate = parseInputValue({ inputValue, customParseDate });
+
+      if (onChange && currentDate && isValid(currentDate) && !isSameDay(value!, currentDate)) {
+        onChange(currentDate);
```
[the remark this answers is not in the document]
Comment (ze-flo) [reply packages/datepickers/src/elements/DatePicker/components/Input.tsx:37]: @jgorfine-zendesk I know this is an edge case, but I think there's still value in testing it. The PR description suggests you might have had a test ready at one point. I'd add it back so a later refactor of this area can't quietly reintroduce the bug.

A good check is to try the test on the unpatched `DatePicker` first and confirm it fails. Then apply the fix and confirm the console event doesn't fire and the test passes.

Here's an AI-generated suggestion:

```tsx
it('does not warn about render-phase updates with a controlled consumer', async () => {
  const consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation(() => {});

  const Controlled = () => {
    const [value, setValue] = useState<Date | undefined>(DEFAULT_DATE);
    return (
      <DatePicker value={value} onChange={setValue}>
        <input data-test-id="input" />
      </DatePicker>
    );
  };

  const { getByTestId, getAllByTestId } = render(<Controlled />);
  const input = getByTestId('input');

  await user.clear(input);
  await user.type(input, '1/4/2019');        // typed-input path
  await user.click(input);
  fireEvent.click(getAllByTestId('day')[1]); // calendar path

  expect(consoleErrorSpy).not.toHaveBeenCalledWith(
    expect.stringContaining('Cannot update a component while rendering'),
    expect.anything()
  );
  consoleErrorSpy.mockRestore();
});
```

[the remark this answers is not in the document]
Comment (jgorfine-zendesk) [reply packages/datepickers/src/elements/DatePicker/components/Input.tsx:37]: Makes sense! Appreciate your take on it; I was honestly on the fence about whether or not to add one. I'll get that going now.

**Review by ze-flo — APPROVED** [review]
Nice! About this [comment](https://github.com/zendeskgarden/react-components/pull/2144/changes#r3826389419) - this edge case is associated with `date-fns` default behavior. I don't think it's completely relevant but I'll let you pick.

**Comment by jgorfine-zendesk** [comment]
Thanks, @ze-flo. I'm currently working on changes that relate to date validity...so I'm going to pass on that suggestion, for now. I'll be sure to keep it in mind going forward.
